### PR TITLE
add timestamp in JOB/PROC STATE print msgs

### PR DESCRIPTION
**Pull request by naughtont3**
Signed-off-by: Thomas Naughton <naughtont@ornl.gov>

### CODE REVIEW

**Comment by naughtont3** [comment]
This adds a timestamp in the state machine transitions that can be useful for tracking time in transitions (where time is spent) from log output. 
 - Example: [prte-state-timestamp.txt](https://github.com/openpmix/prrte/files/4717991/prte-state-timestamp.txt)

**Comment by jjhursey** [comment]
This looks ok to me. I have two notes though:
 * Are we concerned at all about the overhead of calling `gettimeofday` for each of these calls? In the current form that cost is paid on each call even if the message is not printed due to a verbose level below 1 (default). Maybe we can add protection around the call for the normal case?
 * I agree that the timing information is useful. do we need it on the line that follows as well - or it probably doesn't matter since it is likely printed immediately afterward?
```
[node0:16931] [[11879,0],0] [1591108162.118743] ACTIVATE JOB [11879,2] STATE NORMALLY TERMINATED AT prted/pmix/pmix_server_gen.c:463
[node0:16931] [[11879,0],0] ACTIVATING JOB [11879,2] STATE NORMALLY TERMINATED PRI 4    <<<---- Do you need the timestamp here as well
[node0:16931] [[11879,0],0] [1591108162.196824] ACTIVATE PROC [[11879,1],0] STATE NORMALLY TERMINATED AT prted/pmix/pmix_server_gen.c:162
[node0:16931] [[11879,0],0] ACTIVATING PROC [[11879,1],0] STATE NORMALLY TERMINATED PRI 4    <<<---- Do you need the timestamp here as well
```

File: src/mca/state/state.h [hunk 1]
```diff
@@ -74,12 +80,24 @@ PRTE_EXPORT extern prte_mca_base_framework_t prte_state_base_framework;
...
+        struct timeval tv;                                                  \
+        gettimeofday(&tv, NULL);                                            \
+        t  = tv.tv_sec;                                                     \
+        t += (double)tv.tv_usec / 1000000.0;                                \
+    } while(0);
+
 #define PRTE_ACTIVATE_JOB_STATE(j, s)                                       \
     do {                                                                    \
         prte_job_t *shadow=(j);                                             \
+        double timestamp = 0.0;                                             \
```
Comment (jjhursey) [inline src/mca/state/state.h:95]: What you do think about adding a wrapper like this around the print block:
```
if( prte_state_base_framework.framework_verbose > 0 ) {
```

Then we don't incur the overhead of the `gettimeofday` on the calls when verbosity is not turned up

**Comment by naughtont3** [comment]
> This looks ok to me. I have two notes though:
> 
>     * Are we concerned at all about the overhead of calling `gettimeofday` for each of these calls? In the current form that cost is paid on each call even if the message is not printed due to a verbose level below 1 (default). Maybe we can add protection around the call for the normal case?
> 

Yes, I thought about that a little and considered adding configure flag and guarding only use that to select this modified version of macro.  I also think your subsequent suggestion of only calling when verbosity is turned on is better.  I will certainly add verbosity guard.  Do you think we also need configury to avoid `gettimeofday` usage entirely unless asked for?


>     * I agree that the timing information is useful. do we need it on the line that follows as well - or it probably doesn't matter since it is likely printed immediately afterward?
> 

Initially I started to add it there too, but those are not using a macro, just inline prints.  I would prefer to make them use a macro and do similar method to avoid lots of ifdef-ing.  I wanted to go with minimum to start to get comments.  I'm not sure what the priority (PRI) part is indicating here, so might be good to have better insight into why we also show the "ACTIVATING" version (i.e., why are the additional lines there at all?).


> 
> ```
> [node0:16931] [[11879,0],0] [1591108162.118743] ACTIVATE JOB [11879,2] STATE NORMALLY TERMINATED AT prted/pmix/pmix_server_gen.c:463
> [node0:16931] [[11879,0],0] ACTIVATING JOB [11879,2] STATE NORMALLY TERMINATED PRI 4    <<<---- Do you need the timestamp here as well
> [node0:16931] [[11879,0],0] [1591108162.196824] ACTIVATE PROC [[11879,1],0] STATE NORMALLY TERMINATED AT prted/pmix/pmix_server_gen.c:162
> [node0:16931] [[11879,0],0] ACTIVATING PROC [[11879,1],0] STATE NORMALLY TERMINATED PRI 4    <<<---- Do you need the timestamp here as well
> ```

File: src/mca/state/state.h (hunk 1 above)
Comment (naughtont3) [reply src/mca/state/state.h]: updated with this change

**Comment by rhc54** [comment]
> why we also show the "ACTIVATING" version (i.e., why are the additional lines there at all?

We can definitely remove the "PRI 4" part of the second message - it isn't really relevant any more.

There are two lines because the first one is issued at the point where the "activate state" macro is used. However, this just pushes the state into the event queue - it doesn't get executed until the progress thread actually activates the event (which is when the second message comes out).

In most cases, the two steps will happen almost immediately. However, if the daemon is really busy (e.g., in a persistent DVM being driven by a workflow manager that is injecting lots of new jobs), you can begin to see a delay as other events get processed before the new one.

So it just depends on how busy the daemon is, and what you are trying to measure with your time stamp.

File: src/mca/state/state.h (hunk 1 above)
Comment (jjhursey) [reply src/mca/state/state.h]: Thanks!

**Comment by naughtont3** [comment]
@rhc54 OK, so it sounds like it would certainly be good to have the "2nd half" (ACTIVATING) verison too.  I can make that another macro and insert it in all the locations where we currently show the "ACTIVATING" side.  Is that ok to do in this PR?

**Pull request merged**